crates/re_types_builder/src/codegen/cpp.rs

```diff
@@ -167,3 +191,609 @@ fn pragma_once() -> TokenStream {
+                    let comment = comment("Plain Old Data (POD): requires no destructor");
+                    quote! {
+                        case detail::#tag_typename::#tag_ident: {
+                            break; #comment
+                        }
+                    }
+                } else if let Type::Array { elem_type, length } = &obj_field.typ {
+                    // We need special casing for destroying arrays in C++:
+                    let elem_type = quote_element_type(&mut hpp_includes, elem_type);
+                    let length = proc_macro2::Literal::usize_unsuffixed(*length);
+                    quote! {
+                        case detail::#tag_typename::#tag_ident: {
+                            typedef #elem_type TypeAlias;
+                            for (size_t i = #length; i > 0; i -= 1) {
+                                _data.#field_ident[i-1].~TypeAlias();
+                            }
+                            break;
+                        }
+                    }
+                } else {
+                    let typedef_declaration = quote_declaration(
+                        &mut hpp_includes,
+                        obj_field,
+                        &format_ident!("TypeAlias"),
+                        false,
+                    )
+                    .0;
+                    hpp_includes.system.insert("utility".to_owned()); // std::move
+                    quote! {
+                        case detail::#tag_typename::#tag_ident: {
+                            typedef #typedef_declaration;
+                            _data.#field_ident.~TypeAlias();
+                            break;
+                        }
+                    }
+                }
+            }))
+            .collect_vec();
+
+            quote! {
+                ~#pascal_case_ident() {
+                    switch (this->_tag) {
+                        #(#destructor_match_arms)*
+                    }
+                }
+            }
+        };
+
+        let hpp = quote! {
+            #hpp_includes
+
+            namespace rr {
+                namespace #namespace_ident {
+                    namespace detail {
+                        enum class #tag_typename {
+                            #(#tag_fields)*
+                        };
+
+                        union #data_typename {
+                            #(#enum_data_declarations;)*
+
+                            #data_typename() { } // Required by static constructors
+                            ~#data_typename() {}
+
+                            void swap(#data_typename& other) noexcept {
+                                char temp[sizeof(#data_typename)];
+                                std::memcpy(temp, this, sizeof(#data_typename));
+                                std::memcpy(this, &other, sizeof(#data_typename));
+                                std::memcpy(&other, temp, sizeof(#data_typename));
+                            }
```

Comment:
this safes a bunch of switch logic on the tag, but it would be more correct (and safe) to perform a move on the valid fields.
This can ofc only be done inside `struct #pascal_case_ident` meaning we need to have some more logic in there, i.e. more code overall.
The problem with this approach is that we pretend to move in `struct #pascal_case_ident` but don't correctly propagate the move down to the "active field".

The concern is rather academic though. Inclined to keep this as is indefinitely.

Reply:
This code would only break for self-referential structs (as far as I can see), and yeah, we don't ever have those.